lib/s3-compat.js

```diff
@@ -0,0 +1,2051 @@
+                header: header,
+                value: value
+            }, 's3ConditionalHeaders: mapped conditional header');
+        }
+    });
+
+    req.log.debug({
+        conditions: req.conditions
+    }, 's3ConditionalHeaders: final conditions object');
+
+    next();
+}
+
+/*
+ * s3RoleTranslator(req, res, next)
+ * Converts S3 ACL headers to Manta role-tag headers
+ * Arguments:
+ *     req:    HTTP request object containing S3 ACL headers.
+ *     res:    HTTP response object.
+ *     next:   Middleware next function.
+ * Returns:
+ *     None
+ * Notes:
+ *    This function adds the following properties to req.
+ *    - _s3AclOperation : true if we are processing an ACL operation.
+ *    - _s3AclRoles     : Array of Manta Roles translated from S3 ACL.
+ *    - role-tag        : Comma separated string of Manta roles.
+ *
+ * Canned ACLs:
+ * The value of the x-amz-acl header has the name of the canned ACL to enforce
+ * on the resource. Today we only support the following canned ACLs that are
+ * mapped to Manta roles that have the same name,through the S3_ACL_TO_MANTA_-
+ * ROLES object .
+ * +-------------------+--------------+-------------------------------+
+ * | Canned ACL        | Applies to   | Permissions                   |
+ * +-------------------+--------------+-------------------------------+
+ * | private           | Bucket,Object| Owner FULL_CONTROL only.      |
+ * | public-read       | Bucket,Object| Owner FULL_CONTROL, AllUsers  |
+ * |                   |              | READ.                         |
+ * +-------------------+--------------+-------------------------------+
+ *
+ * Grant Headers:
+ *  if any of this headers comes, then parseS3GranToRoles will return
+ *  the appropiate role.
+ *  +-------------------+----------------+----------------------+--------------+
+ *  | Header            | Purpose        | Scope                | Manta Role   |
+ *  +-------------------+----------------+----------------------+--------------+
+ *  | x-amz-grant-read  | Grant READ     | Buckets: list objs.  | public-reader|
+ *  |                   | access         | Objects: read data & | auth-reader  |
+ *  |                   |                | ACL                  |              |
+ *  +-------------------+----------------+----------------------+--------------+
+ *  | x-amz-grant-write | Grant WRITE    | Buckets only: upload | public-writer|
+ *  |                   | access         | or delete objects.   | auth-writer  |
+ *  +-------------------+----------------+----------------------+--------------+
+ *  | x-amz-grant-full- | Grant FULL_    | Buckets: list, write,| public-reader|
+ *  | control           | CONTROL (all   | ACL mgmt. Objects:   | public-writer|
+ *  |                   | perms)         | read, write, ACL mgmt| auth-reader  |
+ *  |                   |                |                        auth-writer  |
+ *  +-------------------+----------------+----------------------+--------------+
```

Comment:
At first glance it feels like manta roles are more fine-grained, but maybe AWS slices differently and we have to meet in the middle?